dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java

```diff
@@ -505,223 +463,703 @@ private String buildTrackedEntityInstanceHql( TrackedEntityInstanceQueryParams p
+            .append( getFromSubQueryOrderAttributes( params ) )
+            .toString();
+    }
 
-        sql += addWhereConditionally( hlp, params.hasTrackedEntityInstances(),
-            () -> " tei.uid in (" + getQuotedCommaDelimitedString( params.getTrackedEntityInstanceUids() ) + ")" );
+    private String getFromSubQueryOrderAttributes( TrackedEntityInstanceQueryParams params )
+    {
+        StringBuilder orderAttributes = new StringBuilder();
 
-        if ( !params.hasTrackedEntityType() )
+        for ( QueryItem orderAttribute : getOrderAttributes( params ) )
         {
-            sql += hlp.whereAnd() + " tei.trackedentitytypeid in (" + params.getTrackedEntityTypes().stream()
-                .filter( Objects::nonNull )
-                .map( TrackedEntityType::getId )
-                .map( String::valueOf )
-                .collect( Collectors.joining( ", " ) ) + ") ";
+            orderAttributes
+                .append( ", " )
+                .append( statementBuilder.columnQuote( orderAttribute.getItemId() ) )
+                .append( ".value AS " )
+                .append( statementBuilder.columnQuote( orderAttribute.getItemId() ) );
+        }
+
+        return orderAttributes.toString();
+    }
+
+    private String getFromSubQueryTrackedEntityConditions( SqlHelper whereAnd, TrackedEntityInstanceQueryParams params )
+    {
+        StringBuilder trackedEntity = new StringBuilder();
+
+        if ( params.hasTrackedEntityType() )
+        {
+            trackedEntity
+                .append( whereAnd.whereAnd() )
+                .append( "TEI.trackedentitytypeid = " )
+                .append( params.getTrackedEntityType().getId() )
+                .append( SPACE );
         }
         else
         {
-            sql += hlp.whereAnd() + " tei.trackedentitytypeid = " + params.getTrackedEntityType().getId() + " ";
+            trackedEntity
+                .append( whereAnd.whereAnd() )
+                .append( "TEI.trackedentitytypeid IN (" )
+                .append( getCommaDelimitedString( getIdentifiers( params.getTrackedEntityTypes() ) ) )
+                .append( ") " );
         }
 
-        params.handleOrganisationUnits();
+        if ( params.hasTrackedEntityInstances() )
+        {
+            trackedEntity
+                .append( whereAnd.whereAnd() )
+                .append( "TEI.uid IN (" )
+                .append( getQuotedCommaDelimitedString( params.getTrackedEntityInstanceUids() ) )
+                .append( ") " );
+        }
 
-        if ( params.isOrganisationUnitMode( OrganisationUnitSelectionMode.ALL ) )
+        if ( params.hasLastUpdatedDuration() )
         {
-            // No restriction
+            trackedEntity.append( whereAnd.whereAnd() )
+                .append( " TEI.lastupdated >= '" )
+                .append( getLongGmtDateString( DateUtils.nowMinusDuration( params.getLastUpdatedDuration() ) ) )
+                .append( SINGLE_QUOTE );
```

Comment:
This part where we append single quotes could be improved to make it more concise, readable and safer.

I have added a method `SqlUtils.singleQuote` which can be used to wrap query arguments instead of  `' + param + '`.

It also fixes a weakness in that if any of the query params contains a single quote, the SQL query will fail and be invalid. Single quotes should be escaped by another single quote (`''`).

Reply:
Agreed. It was more or less a copy paste from 2.34 with the least changes possible to avoid introducing any other accidental bugs. There are several areas of optimisations and cleanup possibilities here.